contract/src/contract.rs

```diff
@@ -95,24 +177,130 @@ pub fn update_config(
+    cw20_msg: Option<Cw20ReceiveMsg>,
+) -> Result<Response, ContractError> {
     //check the only one token is sent
-    if info.funds.len() != 1 {
-        return Err(ContractError::IncorrectTokenQuantity {});
-    }
-
-    //Only burn whale tokens
-    if info.funds[0].denom != "uwhale" {
-        return Err(ContractError::IncorrectToken {});
-    }
     let config: Config = CONFIG.load(deps.storage)?;
+
     let mut messages: Vec<CosmosMsg> = vec![];
-    let amount = info.funds[0].amount;
 
-    //burn the whale sent
-    messages.push(CosmosMsg::from(BankMsg::Burn {
-        amount: info.funds.clone(),
-    }));
+    match config.burn_cw20_addr {
+        Some(burn_cw20_addr) => {
+            // If the BurnCw20_addr was set, we need to not check the 'funds'
+            // We can check the input amount from the cw20_msg
+            // If the BurnCw20_addr is set, we need to check the cw20_msg
+            if cw20_msg.is_none() {
+                return Err(ContractError::IncorrectTokenQuantity {});
+            }
+            let cw20_msg = cw20_msg.clone().unwrap();
+
+            if cw20_msg.amount.is_zero() {
+                return Err(ContractError::IncorrectTokenQuantity {});
+            }
+            messages.push(build_burn_cw20_token_msg(
+                burn_cw20_addr.to_string(),
+                cw20_msg.amount,
+            )?);
+        }
+        None => {
+            if info.funds.len() != 1 {
+                return Err(ContractError::IncorrectTokenQuantity {});
+            }
+            //Only burn whale tokens
+            if info.funds[0].denom != "uwhale" {
+                return Err(ContractError::IncorrectToken {});
+            }
+            messages.push(CosmosMsg::from(BankMsg::Burn {
+                amount: info.funds.clone(),
+            }));
+        }
+    }
+    // Big deviation here lets break it down, end result we want the amount burned and all the messages ready
+    // If using cw20 (first case) we will need to use the CW20 spec and rely on the token address being stored in Config
+    // Otherwise we are proceeding as normal using Native tokens and minting
+    let (amount, ash_amount) = match config.use_cw20 {
+        true => {
+            let provided_amount = cw20_msg.unwrap().amount;
+
+            // But dont return, after this we are going to mint the amount to the sender
+            if config.fee_collector_addr != Addr::unchecked("") {
+                let fee = provided_amount * config.burn_fee;
+
+                messages.push(build_transfer_cw20_token_msg(
+                    config.fee_collector_addr.clone(),
+                    config.mint_denom.clone(),
+                    fee,
+                )?);
+            }
+
+            let ash_amount = coins(provided_amount.u128(), config.mint_denom.as_str());
+
+            // Now mint the ash to the sender, first do a CW20 Mint, this contract already has minter rights
+            messages.push(build_mint_cw20_token_msg(
```

Comment:
why not mint them directly to the sender address?

Reply:
Because the pattern we follow with Native until now was burn -> mint -> transfer and I wanted to keep an identical XP 
I could remove the transfer and then mint to sender but I was thinking do this when it is possible in TF also and do both